### PR TITLE
Small fixes for installing plugin for Firefox on Linux

### LINKED ISSUE #618: Firefox doesn't detect the plugin on Ubuntu
## In order to help us understand your issue, please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your issue (like this: `[x]`)
- Please fill in the sections relevant to your issue and delete the rest.
- Use the *Preview* tab to see what your issue will actually look like

---

### What is the purpose of your issue?
- [ ] Site Support Request: Lightspark does not work correctly on a specific website.
- [x] Bug: Lightspark has the same problem across multiple websites or SWF files.
- [ ] Feature request: a request for new features or support for new platforms, *not sites or SWF files*.
- [ ] Question, Comment, or Suggestion
- [ ] If none of the above apply, check this box.

---

### The following sections should be filled in based on the purpose listed above. Erase any section (the contents between triple ---) that does not apply.

---

#### Provide the complete URL of the page with the issue here:

URL of page

#### If it is possible to download the file and run it with the command line, please provide a link to the file here:

URL of file

#### Provide the complete set of instructions required to reproduce the problem:

1. Load the page.
2. *Insert second action here*
3. *Insert third action, and so on*

#### What kind of failure is your issue?

- [ ] **Fault**: You receive a black window instead of the flash content with a white text such as ``We're sorry, Lightspark encountered an as yet unsupported Flash file``
- [ ] **Crash**: The window disappears or your browser crashes with a message such as ``A tab has crashed``
- [ ] **Broken**: Lightspark shows something, but most of the file is not usable
- [ ] **Glitch**: Lightspark shows something and most of the file is usable, but things are missing or wrong compared to official Flash.
- [ ] **Performance**: Lightspark runs slow or uses lots of memory trying to play the file

#### Provide the Lightspark version

Output of ``lightspark -v``:

```
INFO: Lightspark version 0.8.3-833-gb7395bfd Copyright 2009-2013 Alessandro Pignotti and others
```

#### What operating system are you are running on?

- [ ] Windows 7, 8 or 10
- [ ] Windows XP, Vista or older
- [x] Linux: Ubuntu 18.04
- [ ] Mac OS X: *<specify version here, such as 10.8.1>*
- [ ] Other: *<specify OS and version here>*

#### How are you are running Lightspark?

- [ ] Browser plugin for Chrome or Chromium: *<specify Chromium version here>*
- [ ] Browser plugin for Waterfox: *<specify Waterfox version here>*
- [x] Browser plugin for other browser: Firefox 82.0.2 (64-bit)
- [ ] Lightspark executable on the command line

#### Which of these actions you have taken to help debug the problem?

- [ ] I attached a screen shot of the problem to this issue
- [ ] I have a backtrace listing and pasted it in the bottom section of this issue (*hint: hitting Ctrl-C after a crash will copy it to the clipboard*)
- [ ] I have reproduced the problem using the Lightspark command line executable
- [ ] I ran `lightspark -l 4` on the command line and am attaching logs to this issue (*hint: if they are too big, zip them*)

---

### If this issue is a *feature request*:

Identify the type of request:

- [ ] Support for a new OS: *<name the operating system here>*
- [ ] Support for a new browser: *<name the operating system here>*
- [ ] Support for embedding in an application: *<name the application here>*

---

### If this issue is a *question, comment, or suggestion*:

Identify the subject of this issue:

- [x] The documentation of how to use Lightspark, including the wiki
- [ ] The build instructions to compile Lightspark from source
- [ ] The project and how it is managed (people, process, and so on)
- [ ] Other subject matter not listed

---

### Description of your issue

I installed lightspark using `make install` and it copies the firefox plugin into `/usr/local/lib/mozilla/plugins/liblightsparkplugin.so`.
But lightspark is not listed in Firefox `about:plugins` page. I'm not sure lightspark plugin is installed at the right place since there is nothing else in the folder `/usr/local/lib/mozilla`.
Is there an additional step to install lightspark in Firefox?

### PR DESCRIPTION
**Pull request by PSamuelC**
Previously. `make install` command in build script would put the plugin .so file in a directory where Firefox would not automatically load it. This PR adds some logic to the CMake file to detect if a newer version of Firefox is installed, and changes the plugin target directory accordingly.

Also includes related README update.

Change fixes #618 I think.

### CODE REVIEW

**Comment by PSamuelC** [comment]
Possibly closes issue #618

File: CMakeLists.txt [hunk 1]
```diff
@@ -199,13 +199,23 @@ SET(ENABLE_RTMP TRUE CACHE BOOL "Enable librtmp and dependent functionality?")
...
+# Try to detect where we should install the plugin to, as Firefox version 62+ uses a different directory
+# Unset it first in case we've changed browsers or something between builds
+UNSET(PLUGIN_DIRECTORY CACHE)
+IF(UNIX)
+  EXECUTE_PROCESS(COMMAND firefox --version COMMAND grep -Eo "[0-9]+" COMMAND head -n 1 OUTPUT_VARIABLE FIREFOX_MAJOR_VERSION OUTPUT_STRIP_TRAILING_WHITESPACE)
+  MESSAGE(STATUS "Firefox version ${FIREFOX_MAJOR_VERSION} detected")
+ENDIF()
+IF(FIREFOX_MAJOR_VERSION GREATER 62)
+  SET(PLUGIN_DIRECTORY "$ENV{HOME}/.mozilla/plugins" CACHE STRING "Directory to install Firefox plugin to")
+ELSEIF(ENABLE_DEBIAN_ALTERNATIVES OR WIN33)
```
Comment (dbluelle) [inline CMakeLists.txt:215]: This should be WIN32, I suppose.

**Comment by dbluelle** [comment]
How does this work if you execute "sudo make install". Wouldn't that install the plugin into /root/.mozilla/plugins ?
Also by using this mechanism you have to install the plugin manually for every user...

File: CMakeLists.txt (hunk 1 above)
Comment (PSamuelC) [reply CMakeLists.txt:215]: Not sure how that got in - yes it should be, will change, thanks.

**Comment by PSamuelC** [comment]
> How does this work if you execute "sudo make install"? Wouldn't that install the plugin into /root/.mozilla/plugins ?

Not on Ubuntu, as $HOME is preserved, but it looks like [you might be right for Debian](https://unix.stackexchange.com/questions/91384/how-is-sudo-set-to-not-change-home-in-ubuntu-and-how-to-disable-this-behavior); I'll tinker with this a bit more and see if I can improve that to work on non-Ubuntu distros if that'd be useful.

> Also by using this mechanism you have to install the plugin manually for every user...

That's true I suppose, but I don't think there _is_ a way to put it in one directory and have it loaded it for every user. The /usr/lib/... path doesn't work on newer Firefox (for example see [this article on Debian Wiki](https://wiki.debian.org/FlashPlayer/#Firefox) and the linked issue #618. It's quite possible I'm missing something here however, let me know if there's a folder that would work.

**Comment by PSamuelC** [comment]
Okay, so it looks like /usr/lib/mozilla/plugins *does* get loaded so I've edited it to use that instead of /usr/local/lib/mozilla/plugins and removed the firefox version checking logic for now. If someone confirms that /usr/local... paths are needed on some distro + browser version combination then we can add checks back in.

**Comment by dbluelle** [comment]
I'm still not quite sure that this is the right way to handle the problem.
The whole point of installing everything into /usr/local/ by default is that it doesn't interfere with the normal packaging system of the distribution. Package maintainers will just set the PREFIX to /usr and have everything installed in the right place.
So if you already have lightspark installed from your package manager and then build and install it from source, the plugin will be overwritten and it will not work after you uninstall your selfbuilt lightspark.
I don't know but maybe the browsers have command line parameters or environment variables to set the plugin search path.
But I'm open to suggestions...

**Comment by xvitaly** [comment]
Your PR will break packaged builds, because some distributions don't use `/usr` at all.

You should start using GNUInstallDirs instead of hardcoding any paths.

**Comment by PSamuelC** [comment]
Thanks for the comments all - I'll close the PR

**Comment by dbluelle** [comment]
I'm not that familiar with cmake but I think we need to raise the minimum required cmake version ( to 3.0 ? ) in order to make use of GNUInstallDirs. 
I don't know if older distributions support that cmake version, so we may make building on older distributions impossible only for adjusting some default paths...

**Pull request closed without merging**